File: addons/pos_event/models/pos_order_line.py
from odoo import api, fields, models


class PosOrderLine(models.Model):
    _inherit = "pos.order.line"

    event_ticket_id = fields.Many2one("event.event.ticket")
    event_registration_ids = fields.One2many(
        "event.registration", "pos_order_line_id", string="Event Registrations"
    )

    @api.model
    def _load_pos_data_fields(self, config):
        fields = super()._load_pos_data_fields(config)
        fields += ["event_ticket_id", "event_registration_ids"]
        return fields

File: addons/pos_event/static/src/app/models/pos_order_line.js
// Part of Odoo. See LICENSE file for full copyright and licensing details.
import { PosOrderline } from "@point_of_sale/app/models/pos_order_line";
import { _t } from "@web/core/l10n/translation";
import { patch } from "@web/core/utils/patch";

patch(PosOrderline.prototype, {
    canBeMergedWith(orderline) {
        return (
            this.event_ticket_id?.id === orderline.event_ticket_id?.id &&
            super.canBeMergedWith(...arguments)
        );
    },
    setQuantity(quantity, keep_price) {
        if (this.event_ticket_id && quantity !== "") {
            return {
                title: _t("Ticket error"),
                body: _t("You cannot change quantity for a line linked with an event registration"),
            };
        } else if (this.event_ticket_id) {
            for (const registration of this.event_registration_ids) {
                registration.delete({ silent: true });
            }
        }

        return super.setQuantity(quantity, keep_price);
    },
});
